### ISSUE TITLE
Support for net45 framework

### ISSUE THREAD

**Issue by tameraw** [Azure/azure-iot-sdk-csharp#29]
_From @danstoian on October 3, 2016 5:48_

Hei,

Can you please compile the Microsoft.Azure.Devices package for net45? 
Our project targets net45 (and we cannot change it).

Microsoft.Azure.Amqp requires net451 and I've submitted a pull-request to change the target framework from net451 to net45
(https://github.com/Azure/azure-amqp/issues/39)

Thanks,
Dan

_Copied from original issue: Azure/azure-iot-sdks#853_

**Comment by tameraw** [comment]
@danstoian - Thanks for filing this. We'll look into adding support for .net 4.5 as well.

**Comment by tameraw** [comment]
_From @markrad on October 5, 2016 0:27_

Hi @danstoian,

I reviewed your pull request but you unfortunately made it against the master branch. I can't merge anything into that branch. The develop branch has undergone a significant reorganization so I was having difficulty getting a clean merge of your changes into that branch.

To try and circumvent this, I loaded the develop branch version of iothub_csharp_deviceclient.sln into Visual Studio and modified the project properties of the device project to target .Net 4.5. Unfortunately this caused catastrophic build errors when I attempted a rebuild. I'm still looking into the cause.

_Mark Radbourne MSFT_

**Comment by tameraw** [comment]
_From @danstoian on October 13, 2016 11:53_

Hi @markrad,

I've created a new pull request for the azure-amqp project this time in the develop branch: https://github.com/Azure/azure-amqp/pull/43

I've created a new package for Microsoft.Azure.Amqp locally and then included the new one in iothub_csharp_deviceclient.sln. 

The solution didn't build correctly the first time because there is a new method _void SetMonitor(ITransportMonitor usageMeter)_ in  Microsoft.Azure.Amqp.Transport that need to be implemented in ClientWebSocketTransport and LegacyClientWebSocketTransport... otherwise the solution won't compile.

Since I'm not quite sure what it does I've created an override in both classes that will throw a NotImplementedException.

The bottom line is that the solution compiles locally. I can make a pull-request for _Azure-iot-sdks_ if you wish. You can check my commit at: https://github.com/Azure/azure-iot-sdks/compare/develop...danstoian:develop?expand=1

**Comment by tameraw** [comment]
@danstoian - This is currently WIP. Will update the issue as soon as we have news.

**Comment by tameraw** [comment]
_From @markrad on November 10, 2016 22:29_

Hi @danstoian,

Currently I am pestering the product group to publish your changes to Microsoft.Azure.Amqp. The Nuget package is still targeting .Net 4.5.1. Once they have updated that I will go ahead and make the rest of the changes to target this to .Net 4.5.

_Mark Radbourne MSFT_

**Comment by tameraw** [comment]
_From @danstoian on November 11, 2016 19:49_

Hi @markrad,

Thx for the update. 
I'm looking forward to it.

Br,
Dan

**Comment by tameraw** [comment]
@markrad - A new Nuget package has been released for Microsoft.Azure.Amqp.

**Comment by sergaz-msft** [comment]
Hello @danstoian can you go ahead and try using .net framework 4.5 with the latest version of the Azure IoT sdk? Thanks.

**Comment by danstoian** [comment]
It still doesn´t work. I do get this error when referencing Microsoft.Azure.Devices with a net45 target:

Error		Package Microsoft.Azure.Devices 1.2.4 is not compatible with net45 (.NETFramework,Version=v4.5). Package Microsoft.Azure.Devices 1.2.4 supports:
  - net451 (.NETFramework,Version=v4.5.1)
  - uap10.0 (UAP,Version=v10.0)	

This is how my .csproj looks like:
<ItemGroup Condition=" '$(TargetFramework)' == 'net45' ">
    <PackageReference Include="Newtonsoft.Json" Version="9.0.1" />
    <PackageReference Include="Microsoft.Azure.Devices.Client" Version="1.2.8" />
    <PackageReference Include="Microsoft.Azure.Devices" Version="1.2.4" />
    <Reference Include="System" />
    <Reference Include="Microsoft.CSharp" />
    <Reference Include="System.ComponentModel.DataAnnotations" />
  </ItemGroup>

**Comment by sergaz-msft** [comment]
Hello @danstoian thanks for the details.
If you compile the service and device sdk for .NET45 you will have the dlls on that version. Change the properties (Target Framework to .NET Framework 4.5) on the following projects:

**Service**
- Microsoft.Azure.Devices
- Microsoft.Devices.Shared

**Client**
- Microsoft.Azure.Devices.Client

On your project, instead of adding the NuGet package can you reference the Dlls for the compiled ones?
Let me know the results.
Thanks!

**Comment by sergaz-msft** [comment]
Hello @danstoian,
Letting you know that the next release of the SDK will include support for .NET45 in the Nuget package.
All Best

**Comment by danstoian** [comment]
Thx a lot. Looking forward.

**Comment by Adrya84** [comment]
Hello,

Latest nuget package Microsoft.Azure.Devices 1.16 still targets .net 4.5.1
Any news in which version is included support in nuget for .net 4.5 ?

Thank you,
Adriana

**Comment by CIPop** [comment]
@Adrya84 we don't support the .NET 4.5 TargetFrameworkMoniker and we have no plan in adding that support. New components (Provisioning) will only support `netstandard2.0` (which, for Framework is .NET 4.6.1+).
We would like to understand more about your scenario. The _runtime_ was deprecated a long time ago: https://blogs.msdn.microsoft.com/dotnet/2015/12/09/support-ending-for-the-net-framework-4-4-5-and-4-5-1/ and 4.x versions cannot coexist. This means that you are most likely running on a newer runtime although your app is targetting a lower TFM. Would it be possible to increase the application's TFM to 4.5.1+?
